File: oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

import java.util.Vector;

public enum MapDirection {
    NORTH,
    EAST,
    SOUTH,
    WEST;

    private static final Vector2d[] directions = {
            new Vector2d(0, 1),
            new Vector2d(1, 0),
            new Vector2d(0, -1),
            new Vector2d(-1, 0)
    };

    private static final String[] names = {"Polnoc", "Wschod", "Poludnie", "Zachod"};

    @Override
    public String toString() {
        return names[this.ordinal()];
    }

    public MapDirection next() {
        return MapDirection.values()[((this.ordinal()+1)%4)];
    }

    public MapDirection previous() {
        int index = this.ordinal()-1;
        if (index < 0) index = 3;
        return MapDirection.values()[index];
    }

    public Vector2d toUnitVector () {
        return directions[this.ordinal()];
    }
}

File: KotlinLab/src/main/kotlin/agh/ics/oop/model/MapDirection.kt
package agh.ics.oop.model

enum class MapDirection {
    NORTH,
    EAST,
    SOUTH,
    WEST;

    companion object {
        private val directions = arrayOf(
            Vector2d(0, 1),
            Vector2d(1, 0),
            Vector2d(0, -1),
            Vector2d(-1, 0)
        )

        private val names = arrayOf("Polnoc", "Wschod", "Poludnie", "Zachod")
    }

    override fun toString(): String {
        return names[this.ordinal]
    }

    fun next(): MapDirection {
        return entries[(this.ordinal + 1) % entries.size]
    }

    fun previous(): MapDirection {
        val index = if (this.ordinal - 1 < 0) entries.size - 1 else this.ordinal - 1
        return entries[index]
    }
}
